shared/js/advanced_telemetry_helper.js

```diff
@@ -0,0 +1,154 @@
+/*
+ * This module is a helper for recording telemetry data in Gecko’s telemetry
+ * system. Currently, only data from certified apps will be recorded.
+ *
+ * Usage Example:
+ *  var ath = new AdvancedTelemetryHelper();
+ *  // DO SOMETHING
+ *  // This creates and increments the counter on this metric called
+ *  // 'mycountmetric'
+ *  ath.count('mycountmetric'); // Count will be 1
+ *  ath.count('mycountmetric'); // Count is now 2
+ *  // DO SOMETHING
+ *  // This creates and logs a linear histogram called 'mylinearmetric' with
+ *  // value 100, minimum of 0, max of 1000, with 10 buckets.
+ *  ath.linear('mylinearmetric', 100, 0, 1000, 10);
+ *  // DO SOMETHING
+ *  // This adds to the existing histogram a value of 800.  min/max/buckets
+ *  // are not required on subsequent calls.
+ *  ath.linear('mylinearmetric', 800);
+ *  // DO SOMETHING
+ *  // This creates and logs an exponential histogram called
+ *  // 'myexpmetric' with value 10, minimum of 0, max of 10000, with 10 buckets.
+ *  ath.exp('myexpmetric', 10, 0, 10000, 10);
+ *  // DO SOMETHING
+ *  // This adds to the existing histogram a value of 800.  min/max/buckets
+ *  // are not required on subsequent calls.
+ *  ath.exp('mylinearmetric', 8000);
+ *
+ */
+
+/* globals SettingsListener */
+(function(exports) {
+  'use strict';
+
+  const APP_TELEMETRY_LOG_PREFIX = 'telemetry';
+  const ATH = AdvancedTelemetryHelper;
+  // This is exposed for unit testing purposes
+  ATH.TELEMETRY_ENABLED = false;
+  ATH.TELEMETRY_ENABLED_KEY = 'debug.performance_data.advanced_telemetry';
+  ATH.SETTINGS_LISTENER = false;
+
+  function AdvancedTelemetryHelper() {
+    AdvancedTelemetryHelper.prototype.getTelemetrySetting();
+  }
+
+  AdvancedTelemetryHelper.prototype.getTelemetrySetting = function() {
+    if (ATH.SETTINGS_LISTENER) {
+      return;
+    }
+
+    this.telemetryEnabledListener = function telemetryEnabledListener(enabled) {
+      if (enabled) {
+        ATH.TELEMETRY_ENABLED = true;
+      } else {
+        ATH.TELEMETRY_ENABLED = false;
+      }
+    }.bind(this);
+
+    SettingsListener.observe(ATH.TELEMETRY_ENABLED_KEY,
+      false, this.telemetryEnabledListener);
+    ATH.SETTINGS_LISTENER = true;
+  };
+
+  // API for count histogram entry.
+  // <name> is the unique name of the metric.
+  // This function increments the count for a histogram of type count.
+  AdvancedTelemetryHelper.prototype.count = function(name) {
+    if (!ATH.TELEMETRY_ENABLED) {
+      return;
+    }
+
+    var message;
+ 
+    if (name) {
+      message = APP_TELEMETRY_LOG_PREFIX;
+
+      message += '|' + name;
+      // For a counter histogram, the value makes no difference, just adding
+      // 1 here to keep the parsing consistent.  The last parameter is the
+      // constant Histogram type -- in this case HISTOGRAM_COUNT = 4
+      message += '|1|4';
```

Comment:
I'm curious why the histogram type is added here but is not added for 'linear' or 'exp' histograms.